File: crates/orbit-web/assets/dashboard/reliability.js
// Orbit dashboard pipeline-reliability rendering [ORB-10588].
// Pure vanilla JS, split into ES modules with no build step.
//
// Renders `GET /api/metrics/reliability`: job-run failure rate and recovery
// invocation rate, both computed server-side from durable `job_runs` /
// `invocations` rows. Nothing here derives from tokens or cost.
//
// Two display rules are load-bearing and apply to every rate on this view:
//
//   1. A percentage is never shown without its denominator and its window.
//      `fmtRate` renders "12.5% (n=48)" and the panel header states the
//      window, so a number can't be read out of context.
//   2. A rate whose denominator is below the server's confidence threshold
//      (`low_sample`, set in orbit-core) is withheld — the cell shows the raw
//      counts and an "n too small" marker instead of a percentage.
//
// The server also reports which run states it counted. `succeeded + failed`
// is smaller than the run total whenever runs were cancelled, skipped, or are
// still in flight, so the summary spells out the excluded bucket rather than
// letting the two visible numbers imply they add up.

import { el, syncNodes, fetchJson, getWindow, payloadHonorsWindow, reliabilityWindowFor, wireWindowSelector, syncWindowSelectors } from './common.js';

const $ = (id) => document.getElementById(id);

// Mirrors the windows the endpoint accepts. `all` is deliberately absent:
// a failure rate with no time range is not actionable. When the dashboard
// window is `all`, Reliability keeps a labeled independent 7d cutoff.
const RELIABILITY_WINDOWS = ["1h", "24h", "7d", "30d"];
const DEFAULT_RELIABILITY_WINDOW = "7d";

function getReliabilityWindow() {
  return reliabilityWindowFor(getWindow()).window;
}

const pctFormatter = new Intl.NumberFormat("en-US", {
  style: "percent",
  maximumFractionDigits: 1,
});

function num(value) {
  return typeof value === "number" && Number.isFinite(value) ? value : 0;
}

/// Renders a `Rate` from the API. Returns display text plus the CSS modifier
/// the caller applies, so withheld and low-confidence cells are visually
/// distinct from a real reading.
function describeRate(rate) {
  if (!rate || num(rate.denominator) === 0) {
    return {
      text: "n/a",
      cls: "rel-rate rel-rate-empty",
      title: rate
        ? `no ${rate.denominator_label} in this window`
        : "no data",
    };
  }
  const n = num(rate.denominator);
  const detail = `${num(rate.numerator)} / ${n} ${rate.denominator_label}`;
  if (rate.low_sample) {
    // Withheld, not rounded: a 1-in-3 sample rendered as "33%" reads far more
    // confident than the evidence supports.
    return {
      text: `${num(rate.numerator)}/${n} · n too small`,
      cls: "rel-rate rel-rate-low",
      title: `${detail} — below the confidence threshold, percentage withheld`,
    };
  }
  return {
    text: `${pctFormatter.format(num(rate.value))} (n=${n})`,
    cls: "rel-rate",
    title: detail,
  };
}

function rateCell(rate, extraClass = "") {
  const described = describeRate(rate);
  const node = el("span", {
    class: `${described.cls} ${extraClass}`.trim(),
    text: described.text,
    title: described.title,
  });
  return node;
}

function fmtWindowRange(window) {
  if (!window) return "";
  const since = new Date(window.since);
  const until = new Date(window.until);
  if (Number.isNaN(since.getTime()) || Number.isNaN(until.getTime())) return "";
  const opts = { month: "short", day: "numeric", hour: "2-digit", minute: "2-digit" };
  return `${since.toLocaleString(undefined, opts)} → ${until.toLocaleString(undefined, opts)} UTC${
    window.bucket ? ` · ${window.bucket} buckets` : ""
  }`;
}

function statTile(label, value, hint) {
  return el("div", { class: "rel-tile" }, [
    el("span", { class: "rel-tile-label", text: label }),
    typeof value === "string" ? el("span", { class: "rel-tile-value", text: value }) : value,
    hint ? el("span", { class: "rel-tile-hint", text: hint }) : null,
  ]);
}

/// The headline block: both rates, each with its `n`, plus the explicit
/// statement of what the failure-rate denominator excludes.
function renderSummary(payload) {
  const host = $("reliability-summary");
  if (!host) return;
  const totals = payload.totals || {};
  const runs = totals.job_runs || {};
  const counts = runs.counts || {};
  const recovery = totals.recovery || {};

  const settled = num(counts.succeeded) + num(counts.failed);
  const excluded =
    num(counts.cancelled) + num(counts.skipped) + num(counts.in_flight) + num(counts.unknown);

  const tiles = [
    statTile(
      "Job-run failure rate",
      rateCell(runs.failure_rate),
      `${num(counts.failed)} failed of ${settled} settled runs`,
    ),
    statTile(
      "Recovery per step invocation",
      rateCell(recovery.per_step_invocation),
      recovery.per_step_invocation
        ? `denominator: ${recovery.per_step_invocation.denominator_label}`
        : "",
    ),
    statTile(
      "Job runs needing recovery",
      rateCell(recovery.per_job_run),
      recovery.per_job_run
        ? `denominator: ${recovery.per_job_run.denominator_label}`
        : "",
    ),
    statTile(
      "Runs in window",
      String(num(counts.total)),
      `${settled} settled · ${excluded} not counted`,
    ),
  ];

  const nodes = tiles.map((tile, index) => {
    tile.dataset.key = `rel-tile-${index}`;
    tile.dataset.hash = tile.textContent;
    return tile;
  });
  syncNodes(host, nodes);

  const note = $("reliability-denominator-note");
  if (note) {
    // States that are terminal-but-not-an-outcome, or not terminal at all, sit
    // outside the rate. Saying so here is what keeps "success + failed" from
    // reading as the whole population.
    note.textContent =
      excluded > 0
        ? `Failure rate is over settled runs only (success + failed = ${settled}). ` +
          `${excluded} further run(s) in this window were cancelled, skipped, still in flight, ` +
          `or in an unrecognized state and are excluded from the denominator.`
        : `Failure rate is over settled runs only (success + failed = ${settled}). ` +
          `No cancelled, skipped, in-flight, or unrecognized runs in this window.`;
  }

  const truncNote = $("reliability-truncation-note");
  if (truncNote) {
    truncNote.style.display = runs.truncated ? "" : "none";
    truncNote.textContent = runs.truncated
      ? "Window exceeded the per-read row cap; older runs in this window are not included in these counts."
      : "";
  }
}

function tableFrom(columns, rows, keyFn) {
  const table = el("table", { class: "scoreboard-table rel-table" });
  const thead = el("thead");
  const headRow = el("tr");
  for (const col of columns) {
    headRow.appendChild(el("th", { class: col.num ? "num" : "", text: col.label }));
  }
  thead.appendChild(headRow);
  table.appendChild(thead);

  const tbody = el("tbody");
  for (const row of rows) {
    const tr = el("tr");
    for (const col of columns) {
      const td = el("td", { class: col.num ? "num" : "" });
      const rendered = col.render(row);
      if (typeof rendered === "string") {
        td.textContent = rendered;
      } else if (rendered) {
        td.appendChild(rendered);
      }
      tr.appendChild(td);
    }
    tr.dataset.key = keyFn(row);
    tr.dataset.hash = tr.textContent;
    tbody.appendChild(tr);
  }
  table.appendChild(tbody);
  return table;
}

function emptyState(text) {
  const node = el("div", { class: "empty-state" }, [
    el("div", { class: "icon", text: "✧" }),
    el("div", { class: "text", text }),
  ]);
  node.dataset.key = "empty";
  node.dataset.hash = text;
  return node;
}

/// Failure rate broken out by workspace and by job — the two axes that make a
/// spike attributable rather than merely alarming.
function renderBreakdown(payload) {
  const host = $("reliability-breakdown");
  if (!host) return;
  const workspaces = payload.workspaces || [];
  if (workspaces.length === 0) {
    syncNodes(host, [emptyState("No job runs in this window.")]);
    return;
  }

  const cards = [];
  const wsRows = workspaces.map((ws) => ({
    key: ws.workspace_id,
    label: ws.workspace_name || ws.workspace_id,
    counts: (ws.job_runs || {}).counts || {},
    rate: (ws.job_runs || {}).failure_rate,
  }));
  cards.push(
    card("Failure rate by workspace", tableFrom(
      [
        { label: "workspace", render: (r) => r.label },
        { label: "runs", num: true, render: (r) => String(num(r.counts.total)) },
        { label: "settled", num: true, render: (r) => String(num(r.counts.succeeded) + num(r.counts.failed)) },
        { label: "failed", num: true, render: (r) => String(num(r.counts.failed)) },
        { label: "failure rate", num: true, render: (r) => rateCell(r.rate) },
      ],
      wsRows,
      (r) => `ws-${r.key}`,
    ), "rel-card-workspaces"),
  );

  // Jobs are flattened across workspaces and tagged, so the worst job leads
  // regardless of which workspace it belongs to.
  const jobRows = [];
  for (const ws of workspaces) {
    for (const job of (ws.job_runs || {}).by_job || []) {
      jobRows.push({
        key: `${ws.workspace_id}::${job.job_id}`,
        job_id: job.job_id,
        workspace: ws.workspace_name || ws.workspace_id,
        counts: job.counts || {},
        rate: job.failure_rate,
      });
    }
  }
  jobRows.sort((a, b) => num(b.counts.failed) - num(a.counts.failed) || num(b.counts.total) - num(a.counts.total));
  cards.push(
    card("Failure rate by job", jobRows.length
      ? tableFrom(
          [
            { label: "job", render: (r) => r.job_id },
            { label: "workspace", render: (r) => r.workspace },
            { label: "runs", num: true, render: (r) => String(num(r.counts.total)) },
            { label: "failed", num: true, render: (r) => String(num(r.counts.failed)) },
            { label: "failure rate", num: true, render: (r) => rateCell(r.rate) },
          ],
          jobRows,
          (r) => `job-${r.key}`,
        )
      : emptyState("No job runs in this window."),
      "rel-card-jobs"),
  );

  syncNodes(host, cards);
}

function card(title, body, key) {
  const node = el("div", { class: "audit-summary-card rel-card" }, [
    el("div", { class: "card-title", text: title }),
    el("div", { class: "card-body" }, [body]),
  ]);
  node.dataset.key = key;
  node.dataset.hash = node.textContent;
  return node;
}

/// The over-time series. Each bar carries its own `n` in the tooltip, and a
/// bucket whose denominator is too small is drawn as a muted "low n" bar
/// rather than as a tall confident spike.
function renderOverTime(payload) {
  const host = $("reliability-over-time");
  if (!host) return;
  const workspaces = payload.workspaces || [];

  // Sum buckets across workspaces by start time so the series is machine-wide.
  const merged = new Map();
  for (const ws of workspaces) {
    for (const bucket of (ws.job_runs || {}).over_time || []) {
      const existing = merged.get(bucket.bucket_start) || {
        bucket_start: bucket.bucket_start,
        succeeded: 0,
        failed: 0,
        total: 0,
      };
      existing.succeeded += num((bucket.counts || {}).succeeded);
      existing.failed += num((bucket.counts || {}).failed);
      existing.total += num((bucket.counts || {}).total);
      merged.set(bucket.bucket_start, existing);
    }
  }
  const series = Array.from(merged.values()).sort((a, b) =>
    a.bucket_start < b.bucket_start ? -1 : a.bucket_start > b.bucket_start ? 1 : 0,
  );

  if (series.length === 0) {
    syncNodes(host, [emptyState("No job runs in this window.")]);
    return;
  }

  const bars = series.map((point, index) => {
    const settled = point.succeeded + point.failed;
    const rate = settled > 0 ? point.failed / settled : null;
    const lowSample = settled > 0 && settled < 20;
    const bar = el("div", { class: "rel-bar-slot" });
    const fill = el("div", {
      class: `rel-bar${rate == null ? " rel-bar-empty" : lowSample ? " rel-bar-low" : ""}`,
    });
    // A bucket with no settled runs gets a hairline, not a zero-height bar
    // that would read identically to a clean 0% bucket.
    fill.style.height = rate == null ? "2px" : `${Math.max(2, rate * 100)}%`;
    bar.appendChild(fill);
    bar.title =
      rate == null
        ? `${new Date(point.bucket_start).toLocaleString()} — no settled runs (n=0)`
        : `${new Date(point.bucket_start).toLocaleString()} — ${pctFormatter.format(rate)} failed` +
          ` (${point.failed}/${settled} settled${lowSample ? ", n too small to be confident" : ""})`;
    bar.dataset.key = `bucket-${point.bucket_start}`;
    bar.dataset.hash = `${point.failed}-${settled}`;
    return bar;
  });

  syncNodes(host, bars);
  const axis = $("reliability-over-time-axis");
  if (axis && series.length > 0) {
    axis.textContent = `${new Date(series[0].bucket_start).toLocaleString()} → ${new Date(
      series[series.length - 1].bucket_start,
    ).toLocaleString()}`;
  }
}

/// Per-activity invocation counts with the role each activity was discovered
/// to play. This is the audit trail for the recovery rate: the numerator's
/// membership is visible, not asserted.
function renderActivities(payload) {
  const host = $("reliability-activities");
  if (!host) return;
  const rows = [];
  for (const ws of payload.workspaces || []) {
    for (const activity of (ws.recovery || {}).by_activity || []) {
      rows.push({
        key: `${ws.workspace_id}::${activity.activity_id}`,
        activity_id: activity.activity_id,
        workspace: ws.workspace_name || ws.workspace_id,
        invocation_count: num(activity.invocation_count),
        job_run_count: num(activity.job_run_count),
        role: activity.role || "unknown",
      });
    }
  }
  rows.sort((a, b) => b.invocation_count - a.invocation_count);

  if (rows.length === 0) {
    syncNodes(host, [emptyState("No invocations recorded in this window.")]);
    return;
  }

  syncNodes(host, [
    card("Invocations by activity and discovered role", tableFrom(
      [
        { label: "activity", render: (r) => r.activity_id },
        { label: "role", render: (r) => el("span", { class: `rel-role rel-role-${r.role}`, text: r.role }) },
        { label: "workspace", render: (r) => r.workspace },
        { label: "invocations", num: true, render: (r) => String(r.invocation_count) },
        { label: "runs touched", num: true, render: (r) => String(r.job_run_count) },
      ],
      rows,
      (r) => `act-${r.key}`,
    ), "rel-card-activities"),
  ]);
}

function renderReliabilityScope(payload) {
  const badge = $("reliability-scope-badge");
  if (!badge) return;
  const scope = payload && payload.scope === "workspace" ? "Workspace" : "Fleet-wide";
  const rel = reliabilityWindowFor(getWindow());
  badge.textContent = rel.independent ? `${scope} · independent ${rel.window}` : scope;
  badge.className = "scope-badge independent";
  badge.hidden = false;
  badge.title = rel.independent
    ? "Reliability is fleet-wide and cannot use the unbounded all window"
    : "Reliability is fleet-wide; the workspace selector does not apply";
}

function renderReliability(payload) {
  if (!payload) return;
  const rel = reliabilityWindowFor(getWindow());
  if (!payloadHonorsWindow(payload, rel.window)) {
    console.error(
      `reliability payload window ${(payload.window || {}).label} rejected under ${rel.window} selection`,
    );
    return;
  }
  syncWindowSelectors();
  renderReliabilityScope(payload);
  const meta = $("reliability-meta");
  if (meta) {
    const range = fmtWindowRange(payload.window);
    const unreadable = (payload.unreadable_workspaces || []).length;
    const fleet = "Fleet-wide";
    const independent = rel.independent ? " · independent window" : "";
    meta.textContent = `${fleet} · ${range}${independent}${unreadable ? ` · ${unreadable} workspace(s) unreadable` : ""}`;
  }
  const count = $("reliability-count");
  if (count) {
    const total = num(((payload.totals || {}).job_runs || {}).counts?.total);
    count.textContent = `${total} runs / ${(payload.window || {}).label || ""}`;
  }
  renderSummary(payload);
  renderOverTime(payload);
  renderBreakdown(payload);
  renderActivities(payload);
}

function fetchAndRenderReliability() {
  const rel = reliabilityWindowFor(getWindow());
  return fetchJson(`/api/metrics/reliability?window=${encodeURIComponent(rel.window)}`)
    .then(renderReliability);
}

// Idempotent attach, matching the scoreboard selector's contract.
function wireReliabilityWindowSelector() {
  wireWindowSelector("reliability-window-selector", { allowAll: false });
}

export {
  renderReliability,
  fetchAndRenderReliability,
  wireReliabilityWindowSelector,
  getReliabilityWindow,
  describeRate,
  RELIABILITY_WINDOWS,
  DEFAULT_RELIABILITY_WINDOW,
};

File: crates/orbit-core/src/metrics/reliability.rs
//! Pipeline-reliability aggregation over durable run state [ORB-10588].
//!
//! Answers two operator questions that were previously only reachable by
//! opening SQLite by hand: how often job runs fail, and how often the recovery
//! path fires.
//!
//! # Inputs
//!
//! Everything here derives from persisted `job_runs` and `invocations` rows.
//! No agent-loop output, no token field, and no cost field is read — see
//! [`orbit_store::contracts::JobRunOutcomeFact`] and
//! [`orbit_store::contracts::ActivityInvocationCount`] for the projected column sets.
//!
//! # Outcome classification
//!
//! `job_runs.state` is not two-valued, so a "failure rate" has to say which
//! states it counts. [`RunOutcome`] partitions the [`JobRunState`] variants:
//!
//! | outcome | states | in failure-rate denominator |
//! |---|---|---|
//! | [`RunOutcome::Succeeded`] | `success` | yes |
//! | [`RunOutcome::Failed`] | `failed`, `timeout`, `interrupted` | yes |
//! | [`RunOutcome::Cancelled`] | `cancelled` | no |
//! | [`RunOutcome::Skipped`] | `skipped` | no |
//! | [`RunOutcome::InFlight`] | `pending`, `running`, `retrying` | no |
//! | [`RunOutcome::Unknown`] | any state this build cannot parse | no |
//!
//! `timeout` and `interrupted` count as failures: both are runs the pipeline
//! intended to finish and did not. `interrupted` specifically means the owning
//! process died without finalizing, which is a reliability event even though
//! the run's checkpoints can seed a resumed follow-up. `cancelled` is a
//! deliberate operator action, so counting it as a failure would make manual
//! intervention look like breakage. `skipped` never ran.
//!
//! Because four of the six outcomes sit outside the denominator,
//! `succeeded + failed` does **not** equal the run count in general. Callers
//! render [`OutcomeCounts::settled`] as the `n` and report the excluded
//! buckets separately; [`OutcomeCounts::total`] is kept so the gap is visible
//! rather than implied away.
//!
//! # Recovery attribution
//!
//! Which activity performs recovery is a property of a workspace's job
//! definitions, never of Orbit, so the recovery activity set is discovered by
//! walking the job catalog ([`JobV2::activity_roles`]) at query time. An
//! activity that a catalog uses in both roles is reported as ambiguous and
//! excluded from the numerator, since its invocations cannot be attributed
//! structurally.

use std::collections::{BTreeMap, BTreeSet};

use chrono::{DateTime, Duration, Utc};
use serde::Serialize;

use orbit_common::OrbitError;
use orbit_store::contracts::{ActivityInvocationCount, InvocationRunCoverage, JobRunOutcomeFact};
use orbit_types::workflow::{JobActivityRoles, JobRunState};

use crate::runtime::OrbitRuntime;

/// Row cap for the per-run fact read backing the failure rate. Reaching it
/// sets [`JobRunReliability::truncated`]; the cap is never applied silently.
const MAX_JOB_RUN_FACTS: usize = 200_000;

/// Denominator below which a rate is flagged [`Rate::low_sample`]. Callers
/// mark or withhold such cells instead of rendering a confident percentage.
pub const MIN_CONFIDENT_SAMPLE: u64 = 20;

/// Bucket width for the over-time breakdown.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum BucketGranularity {
    Hour,
    Day,
}

impl BucketGranularity {
    fn width(self) -> Duration {
        match self {
            Self::Hour => Duration::hours(1),
            Self::Day => Duration::days(1),
        }
    }

    /// Floors `at` to the start of its bucket.
    fn floor(self, at: DateTime<Utc>) -> DateTime<Utc> {
        let seconds = self.width().num_seconds().max(1);
        let epoch = at.timestamp();
        let floored = epoch.div_euclid(seconds) * seconds;
        DateTime::from_timestamp(floored, 0).unwrap_or(at)
    }
}

/// The explicit, half-open window (`since <= t < until`) every rate is scoped
/// to. Carried through to the payload so no rate can be rendered without it.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct ReliabilityWindow {
    /// Caller-supplied label for the window (e.g. `"7d"`), echoed to the UI.
    pub label: String,
    pub since: DateTime<Utc>,
    pub until: DateTime<Utc>,
    pub bucket: BucketGranularity,
}

impl ReliabilityWindow {
    /// Builds a window ending at `until` and spanning `duration`, choosing a
    /// bucket width that keeps the series legible.
    pub fn ending_at(label: impl Into<String>, until: DateTime<Utc>, duration: Duration) -> Self {
        let bucket = if duration <= Duration::days(2) {
            BucketGranularity::Hour
        } else {
            BucketGranularity::Day
        };
        Self {
            label: label.into(),
            since: until - duration,
            until,
            bucket,
        }
    }
}

/// How a persisted `job_runs.state` is counted.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RunOutcome {
    Succeeded,
    Failed,
    Cancelled,
    Skipped,
    InFlight,
    Unknown,
}

impl RunOutcome {
    /// Classifies a raw `job_runs.state` string.
    ///
    /// Parsing is deliberately tolerant: a state written by a newer binary
    /// lands in [`RunOutcome::Unknown`] and stays visible in the counts rather
    /// than being silently folded into success or failure.
    pub fn classify(raw: &str) -> Self {
        let Ok(state) = raw.parse::<JobRunState>() else {
            return Self::Unknown;
        };
        match state {
            JobRunState::Success => Self::Succeeded,
            JobRunState::Failed | JobRunState::Timeout | JobRunState::Interrupted => Self::Failed,
            JobRunState::Cancelled => Self::Cancelled,
            JobRunState::Skipped => Self::Skipped,
            JobRunState::Pending | JobRunState::Running | JobRunState::Retrying => Self::InFlight,
        }
    }
}

/// Run counts for one scope (overall, one job, or one time bucket).
#[derive(Debug, Clone, Default, PartialEq, Eq, Serialize)]
pub struct OutcomeCounts {
    /// Every run in scope, including the ones outside the denominator.
    pub total: u64,
    pub succeeded: u64,
    pub failed: u64,
    pub cancelled: u64,
    pub skipped: u64,
    pub in_flight: u64,
    pub unknown: u64,
}

impl OutcomeCounts {
    fn record(&mut self, outcome: RunOutcome) {
        self.total += 1;
        match outcome {
            RunOutcome::Succeeded => self.succeeded += 1,
            RunOutcome::Failed => self.failed += 1,
            RunOutcome::Cancelled => self.cancelled += 1,
            RunOutcome::Skipped => self.skipped += 1,
            RunOutcome::InFlight => self.in_flight += 1,
            RunOutcome::Unknown => self.unknown += 1,
        }
    }

    /// Runs that reached an outcome the failure rate is defined over.
    ///
    /// This is the rate's `n`. It is smaller than [`OutcomeCounts::total`]
    /// whenever runs were cancelled, skipped, still in flight, or in a state
    /// this build cannot parse.
    pub fn settled(&self) -> u64 {
        self.succeeded + self.failed
    }

    /// Runs excluded from the denominator, for the "not counted" disclosure.
    pub fn excluded(&self) -> u64 {
        self.cancelled + self.skipped + self.in_flight + self.unknown
    }

    /// Failure rate over settled runs, or `None` when nothing settled.
    pub fn failure_rate(&self) -> Rate {
        Rate::new(
            self.failed,
            self.settled(),
            "settled runs (success + failed)",
        )
    }
}

/// A rate that carries its own denominator, so no surface can render the
/// number without the `n` it was computed from.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct Rate {
    pub numerator: u64,
    pub denominator: u64,
    /// Human-readable description of what the denominator counts. Rendered
    /// verbatim by the UI so the meaning never lives only in code.
    pub denominator_label: String,
    /// `numerator / denominator`, or `None` when the denominator is zero.
    pub value: Option<f64>,
    /// True when the denominator is below [`MIN_CONFIDENT_SAMPLE`]. Consumers
    /// mark or withhold the percentage instead of presenting it as confident.
    pub low_sample: bool,
}

impl Rate {
    pub fn new(numerator: u64, denominator: u64, denominator_label: impl Into<String>) -> Self {
        Self {
            numerator,
            denominator,
            denominator_label: denominator_label.into(),
            value: (denominator > 0).then(|| numerator as f64 / denominator as f64),
            low_sample: denominator < MIN_CONFIDENT_SAMPLE,
        }
    }
}

/// Counts for one job id.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct JobOutcomeRow {
    pub job_id: String,
    #[serde(flatten)]
    pub counts: OutcomeCounts,
}

/// Counts for one time bucket.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct BucketOutcomeRow {
    pub bucket_start: DateTime<Utc>,
    pub bucket_end: DateTime<Utc>,
    #[serde(flatten)]
    pub counts: OutcomeCounts,
}

/// Job-run failure rate for one workspace over one window.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct JobRunReliability {
    pub overall: OutcomeCounts,
    /// Descending by run count, so the noisiest job leads.
    pub by_job: Vec<JobOutcomeRow>,
    /// Ascending by bucket start, with empty buckets present so a gap in
    /// activity reads as a gap rather than as continuity.
    pub over_time: Vec<BucketOutcomeRow>,
    /// Every raw `job_runs.state` value observed in the window, with its
    /// count. This is the evidence behind the classification table: an
    /// operator can check what the store actually contains without querying
    /// it, and an unrecognized state shows up here by name.
    pub observed_states: BTreeMap<String, u64>,
    /// True when the window held more runs than the read cap.
    pub truncated: bool,
}

/// Recovery-path engagement for one workspace over one window.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct RecoveryReliability {
    /// Activities the job catalog names as `recovery_activity`, discovered at
    /// query time. Empty when no job declares one — in which case the rate is
    /// not applicable rather than zero.
    pub recovery_activities: Vec<String>,
    /// Activities the catalog uses as both a step target and a recovery hook.
    /// Their invocations are excluded from the numerator because the store
    /// does not record which role a given invocation played.
    pub ambiguous_activities: Vec<String>,
    /// Recovery invocations per invocation of a declared step activity. This
    /// is the "how much extra work does recovery cost" reading.
    pub per_step_invocation: Rate,
    /// Share of job runs that recorded at least one invocation and had at
    /// least one recovery invocation. This is the "how often does a run need
    /// recovery at all" reading.
    pub per_job_run: Rate,
    /// Full per-activity invocation counts for the window, so the two rates
    /// above can be audited against the rows they came from.
    pub by_activity: Vec<ActivityCount>,
}

/// One activity's invocation counts, tagged with its discovered role.
#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
pub struct ActivityCount {
    pub activity_id: String,
    pub invocation_count: u64,
    pub job_run_count: u64,
    pub role: ActivityRole,
}

/// The structural role a job catalog assigns an activity.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "snake_case")]
pub enum ActivityRole {
    /// Named only as a step target.
    Step,
    /// Named only as a `recovery_activity`.
    Recovery,
    /// Named in both roles; not attributable from the store alone.
    Ambiguous,
    /// Invoked but not named by any loaded job definition.
    Unknown,
}

/// Both reliability readings for one workspace, plus the window they share.
#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct RuntimeReliability {
    pub window: ReliabilityWindow,
    pub job_runs: JobRunReliability,
    pub recovery: RecoveryReliability,
}

impl OrbitRuntime {
    /// Computes job-run failure rate and recovery engagement for this
    /// workspace over `window`, entirely from persisted store rows.
    pub fn pipeline_reliability(
        &self,
        window: &ReliabilityWindow,
    ) -> Result<RuntimeReliability, OrbitError> {
        if window.since >= window.until {
            return Err(OrbitError::InvalidInput(
                "reliability window `since` must be earlier than `until`".to_string(),
            ));
        }
        let workspace_id = self.workspace_id()?;
        let store = self.sqlite_store()?;

        let facts = store.list_job_run_outcome_facts(
            &workspace_id,
            window.since,
            window.until,
            MAX_JOB_RUN_FACTS,
        )?;
        let job_runs = aggregate_job_runs(&facts.facts, window, facts.truncated);

        let counts =
            store.count_invocations_by_activity(&workspace_id, window.since, window.until)?;
        let roles = self.catalog_activity_roles()?;
        // The distinct-run denominator has to come from the store: distinct
        // counts do not sum across activities.
        let recovery_ids = roles.recovery_only().into_iter().collect::<Vec<_>>();
        let coverage = store.count_invocation_job_runs(
            &workspace_id,
            window.since,
            window.until,
            &recovery_ids,
        )?;
        let recovery = aggregate_recovery(&counts, &roles, coverage);

        Ok(RuntimeReliability {
            window: window.clone(),
            job_runs,
            recovery,
        })
    }

    /// Unions the activity roles declared by every job in the catalog.
    ///
    /// Disabled jobs are included: a job disabled midway through the window
    /// still produced the invocations being counted.
    fn catalog_activity_roles(&self) -> Result<JobActivityRoles, OrbitError> {
        use crate::application::job::JobCatalogFilter;

        let mut roles = JobActivityRoles::default();
        for (entry, _) in self.list_job_catalog_with_last_run(true, JobCatalogFilter::All)? {
            roles.merge(&entry.spec.activity_roles());
        }
        Ok(roles)
    }
}

/// Buckets and totals the per-run facts. Pure, so it is exercised directly by
/// unit tests without a store.
pub fn aggregate_job_runs(
    facts: &[JobRunOutcomeFact],
    window: &ReliabilityWindow,
    truncated: bool,
) -> JobRunReliability {
    let mut overall = OutcomeCounts::default();
    let mut by_job: BTreeMap<String, OutcomeCounts> = BTreeMap::new();
    let mut by_bucket: BTreeMap<DateTime<Utc>, OutcomeCounts> = BTreeMap::new();
    let mut observed_states: BTreeMap<String, u64> = BTreeMap::new();

    for fact in facts {
        let outcome = RunOutcome::classify(&fact.state);
        overall.record(outcome);
        by_job
            .entry(fact.job_id.clone())
            .or_default()
            .record(outcome);
        by_bucket
            .entry(window.bucket.floor(fact.created_at))
            .or_default()
            .record(outcome);
        *observed_states.entry(fact.state.clone()).or_default() += 1;
    }

    let mut job_rows = by_job
        .into_iter()
        .map(|(job_id, counts)| JobOutcomeRow { job_id, counts })
        .collect::<Vec<_>>();
    job_rows.sort_by(|left, right| {
        right
            .counts
            .total
            .cmp(&left.counts.total)
            .then_with(|| left.job_id.cmp(&right.job_id))
    });

    JobRunReliability {
        overall,
        by_job: job_rows,
        over_time: fill_buckets(window, &by_bucket),
        observed_states,
        truncated,
    }
}

/// Materializes every bucket in the window, including empty ones, so the
/// series cannot imply activity across a quiet stretch.
fn fill_buckets(
    window: &ReliabilityWindow,
    observed: &BTreeMap<DateTime<Utc>, OutcomeCounts>,
) -> Vec<BucketOutcomeRow> {
    let width = window.bucket.width();
    let mut rows = Vec::new();
    let mut start = window.bucket.floor(window.since);
    // A pathological window cannot outrun this: the loop advances by a fixed
    // positive width and stops at `until`.
    while start < window.until {
        let end = start + width;
        rows.push(BucketOutcomeRow {
            bucket_start: start,
            bucket_end: end,
            counts: observed.get(&start).cloned().unwrap_or_default(),
        });
        start = end;
    }
    rows
}

/// Attributes invocation counts to discovered roles and derives both recovery
/// rates. Pure, for the same reason as [`aggregate_job_runs`].
pub fn aggregate_recovery(
    counts: &[ActivityInvocationCount],
    roles: &JobActivityRoles,
    coverage: InvocationRunCoverage,
) -> RecoveryReliability {
    let recovery_only = roles.recovery_only();
    let step_only = roles.step_only();
    let ambiguous = roles
        .recovery
        .intersection(&roles.step)
        .cloned()
        .collect::<BTreeSet<_>>();

    let mut recovery_invocations = 0_u64;
    let mut step_invocations = 0_u64;
    let mut by_activity = Vec::with_capacity(counts.len());

    for count in counts {
        let role = if ambiguous.contains(&count.activity_id) {
            ActivityRole::Ambiguous
        } else if recovery_only.contains(&count.activity_id) {
            ActivityRole::Recovery
        } else if step_only.contains(&count.activity_id) {
            ActivityRole::Step
        } else {
            ActivityRole::Unknown
        };
        match role {
            ActivityRole::Recovery => recovery_invocations += count.invocation_count,
            ActivityRole::Step => step_invocations += count.invocation_count,
            ActivityRole::Ambiguous | ActivityRole::Unknown => {}
        }
        by_activity.push(ActivityCount {
            activity_id: count.activity_id.clone(),
            invocation_count: count.invocation_count,
            job_run_count: count.job_run_count,
            role,
        });
    }

    RecoveryReliability {
        recovery_activities: recovery_only.into_iter().collect(),
        ambiguous_activities: ambiguous.into_iter().collect(),
        per_step_invocation: Rate::new(
            recovery_invocations,
            step_invocations,
            "step-activity invocations",
        ),
        per_job_run: Rate::new(
            coverage.matching_job_runs,
            coverage.total_job_runs,
            "job runs with any recorded invocation",
        ),
        by_activity,
    }
}
